# Proyecto RappiPlus: de datos a decisiones de negocio

**Introducción**


El objetivo de este proyecto es evaluar el desempeño del servicio **RappiPlus** para apoyar **decisiones de negocio basadas en datos**.

Se trabajan con múltiples datasets del negocio:

- **rappiplus_orders_raw.csv** → información de pedidos, precios, descuentos y revenue  
- **rappiplus_catalog.csv** → costos de productos, categorías y proveedores  
- **rappiplus_marketing_spend.csv** → inversión en marketing por canal y país  
- **events / users / user_activity (SQL)** → comportamiento del usuario dentro de la plataforma  
- **experiment_checkout_ui.csv** → resultados de un experimento A/B en el checkout  

El análisis sigue una lógica clara y progresiva:

1. 🔍 Evaluar si podemos confiar en los datos (calidad de datos en Python) 

2. 💰 Analizar si el negocio es rentable (revenue, costos y profit)  

3. 🛒 Entender dónde se pierden los usuarios (funnel de conversión)  

4. 🔁 Evaluar si los usuarios regresan (retención por cohortes)  

5. 🧪 Validar si los cambios generan impacto (test estadístico)  

6. 📊 Comunicar los resultados (dashboard en BI)  

A lo largo del proyecto, se transforman datos en insights para responder preguntas clave del negocio y proponer **recomendaciones accionables**.

---

## 🔹 Paso 1: Cargar y validar la calidad de los datos

---

### 1.1 Carga de datos y vista rápida

**🎯 Objetivo:** Familiarizarte con la estructura de los datasets del negocio antes de analizarlos.

**Instrucciones:**

- Importa las librerías necesarias
- Carga los archivos:
  - `rappiplus_orders_raw.csv`
  - `rappiplus_catalog.csv`
  - `rappiplus_marketing_spend.csv`
- Guarda los DataFrames en:
  - `orders`, `catalog`, `marketing`
- Explora cada dataset.

---

In [85]:
# importar librerías
import pandas as pd

In [86]:
# cargar archivos
orders = pd.read_csv('https://practicum-content.s3.amazonaws.com/datasets/rappiplus_orders_raw.csv')
catalog = pd.read_csv('https://practicum-content.s3.amazonaws.com/datasets/rappiplus_catalog.csv')
marketing = pd.read_csv('https://practicum-content.s3.amazonaws.com/datasets/rappiplus_marketing_spend.csv')

In [87]:
# explorar dataset
orders.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25100 entries, 0 to 25099
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   id_pedido           25100 non-null  object 
 1   id_usuario          25100 non-null  object 
 2   fecha_hora_pedido   25100 non-null  object 
 3   pais                24800 non-null  object 
 4   dispositivo         25080 non-null  object 
 5   fuente_referencia   25070 non-null  object 
 6   nombre_producto     25070 non-null  object 
 7   categoria_producto  25020 non-null  object 
 8   cantidad            25050 non-null  float64
 9   precio_unitario     25050 non-null  float64
 10  monto_descuento     25050 non-null  float64
 11  monto_total         25100 non-null  float64
dtypes: float64(4), object(8)
memory usage: 2.3+ MB


In [88]:
orders.head()  

,id_pedido,id_usuario,fecha_hora_pedido,pais,dispositivo,fuente_referencia,nombre_producto,categoria_producto,cantidad,precio_unitario,monto_descuento,monto_total
0,order_0,user_6993,2025-05-22,Argentina,desktop,organic,Jacket-Winter-M,Moda,2.0,332.69,0.0,665.37
1,order_1,user_1329,2025-06-15,Mexico,desktop,paid_search,Tablet-Standard-64GB,Electronica,1.0,176.86,5.0,171.86
2,order_2,user_3194,2025-05-02,Argentina,desktop,social,Blender-XL-Red,Hogar,2.0,102.99,10.0,195.99
3,order_3,user_4510,2025-06-09,Colombia,mobile,social,Tablet-Standard-64GB,Electronica,1.0,257.87,15.0,242.87
4,order_4,user_5044,2025-03-30,Argentina,desktop,paid_search,Blender-XL-Red,Hogar,1.0,336.28,0.0,336.28


In [89]:
catalog.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7 entries, 0 to 6
Data columns (total 4 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   nombre_producto     7 non-null      object 
 1   categoria_producto  7 non-null      object 
 2   costo_unitario      7 non-null      float64
 3   proveedor           7 non-null      object 
dtypes: float64(1), object(3)
memory usage: 352.0+ bytes


In [90]:
catalog.head(10)

,nombre_producto,categoria_producto,costo_unitario,proveedor
0,Laptop-Gaming-16GB,Electrónica,280.68,"Fuller, Pena and Myers"
1,Phone-Pro-128GB,Electrónica,10.12,King Ltd
2,Tablet-Standard-64GB,Electrónica,25.21,Bowers LLC
3,Blender-XL-Red,Hogar,176.64,Long-Reid
4,Vacuum-Pro-Black,Hogar,16.60,"Rivera, Carr and Finley"
5,Sneakers-Urban-42,Moda,17.21,Greene-Smith
6,Jacket-Winter-M,Moda,189.31,Mcmillan-Rhodes


In [91]:
marketing.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1620 entries, 0 to 1619
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   fecha       1620 non-null   object 
 1   pais        1620 non-null   object 
 2   id_campaña  1620 non-null   object 
 3   canal       1519 non-null   object 
 4   gasto       1620 non-null   float64
dtypes: float64(1), object(4)
memory usage: 63.4+ KB


In [92]:
marketing.head()

,fecha,pais,id_campaña,canal,gasto
0,2025-01-01,Mexico,organic_Mexico,organic,2446.25
1,2025-01-01,Mexico,paid_search_Mexico,paid_search,2704.34
2,2025-01-01,Mexico,social_Mexico,social,2045.01
3,2025-01-01,Colombia,organic_Colombia,organic,2597.21
4,2025-01-01,Colombia,paid_search_Colombia,paid_search,1771.40


**Nota:** se cargan los archivos csv ***rappiplus_orders_raw.csv***, ***rappiplus_catalog.csv***, ***rappiplus_marketing_spend.csv*** para su correspondiente revision, se utiliza ***pandas*** para su correcta manipulación y revisión. Se utilizan los métodos ***head()*** e ***info()*** para explorar rapidamente el tipo de dato que tiene cada columna de la tabla y dar un breve vistazo a la información de los registros de las mismas

Se puede apreciar en la revisión rapida que la mayoria de las columnas tiene su tipo de dato en el formato correcto, solo se presentan las columnas con fechas con un tipo de dato que no es el adecuado, se prentende realizar el cambio del formato object al formato datatime64 para su correcta manipulación. 

De igual forma se aprecian datos registros faltantes en algunas de las columnas son el reporte arrojado por ***.info*** en las columnas de orders (*pais , dispositivo, fuente_referencia, nombre_producto, categoria_producto, cantidad, precio_unitario, monto_descuento*) y marketing (*canal*), mientras que en catalog se aprecia que todos los registros cuentan con datos. Se usaran tecnicas de revisión de datos para corregir y tomar la desición correcta de que hacer con estos faltantes. 

---


### Revisión y calidad de datos

**🎯 Objetivo:** Detectar y corregir problemas en los datos que puedan afectar el análisis de revenue, costos y rentabilidad.

Se revisan lo
s 3 datasets
- Validar y convertir fechas al formato correcto  
- Revisar variables numéricas (sin negativos o ceros inválidos)  
- Verificar consistencia de montos  
- Eliminar duplicados  
- Revisar variables categóricas 

---

In [93]:
orders['fecha_hora_pedido'] = pd.to_datetime(orders['fecha_hora_pedido'])
marketing['fecha'] = pd.to_datetime(marketing['fecha'])

In [94]:
orders.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25100 entries, 0 to 25099
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   id_pedido           25100 non-null  object        
 1   id_usuario          25100 non-null  object        
 2   fecha_hora_pedido   25100 non-null  datetime64[ns]
 3   pais                24800 non-null  object        
 4   dispositivo         25080 non-null  object        
 5   fuente_referencia   25070 non-null  object        
 6   nombre_producto     25070 non-null  object        
 7   categoria_producto  25020 non-null  object        
 8   cantidad            25050 non-null  float64       
 9   precio_unitario     25050 non-null  float64       
 10  monto_descuento     25050 non-null  float64       
 11  monto_total         25100 non-null  float64       
dtypes: datetime64[ns](1), float64(4), object(7)
memory usage: 2.3+ MB


In [95]:
marketing.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1620 entries, 0 to 1619
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype         
---  ------      --------------  -----         
 0   fecha       1620 non-null   datetime64[ns]
 1   pais        1620 non-null   object        
 2   id_campaña  1620 non-null   object        
 3   canal       1519 non-null   object        
 4   gasto       1620 non-null   float64       
dtypes: datetime64[ns](1), float64(1), object(3)
memory usage: 63.4+ KB


**Nota:** Se detecta formato incorrecto en  fechas en el vistazo rapido y se utilizo la herramienta ***pd.to_datetime()*** para realizar su corrección al formato datetime64, el adecuado para manejar formatos de fecha.

In [96]:

def revisar_negativos(datos):
    columnas_numericas = datos.select_dtypes(include=['number']).columns
    for columnas in columnas_numericas:
        negativos = (datos[columnas] <= 0).sum()
        print(f'{columnas}:{negativos}')
    print()

        


print('negativos en orders')
revisar_negativos(orders)
print('negativos catalog')
revisar_negativos(catalog)
print('negativos Marketing')
revisar_negativos(marketing)


negativos en orders
cantidad:4
precio_unitario:0
monto_descuento:12551
monto_total:4

negativos catalog
costo_unitario:0

negativos Marketing
gasto:0



In [97]:
#verificación negativos de monto descuento ya que esta si puede tener 0 pero no negativos
(orders['monto_descuento'] < 0).sum()

0

In [98]:
orders[orders['cantidad']<=0]

,id_pedido,id_usuario,fecha_hora_pedido,pais,dispositivo,fuente_referencia,nombre_producto,categoria_producto,cantidad,precio_unitario,monto_descuento,monto_total
266,order_266,user_7011,2025-03-13,NaN,desktop,paid_search,Phone-Pro-128GB,Electronica,-2.0,101.31,10.0,-192.62
267,order_267,user_1087,2025-05-07,NaN,desktop,social,Phone-Pro-128GB,Electronica,-1.0,43.50,5.0,-38.50
268,order_268,user_84,2025-02-19,NaN,desktop,organic,Phone-Pro-128GB,Electronica,-1.0,497.65,5.0,-492.65
269,order_269,user_3323,2025-05-25,NaN,desktop,paid_search,Phone-Pro-128GB,Electronica,-1.0,423.53,0.0,-423.53


In [99]:
#correccion de negativos 
def correccion_negativos(datos):
    columnas_numericas = datos.select_dtypes(include=['number']).columns
    for columnas in columnas_numericas:
        datos[columnas] = datos[columnas].abs()

correccion_negativos(orders)
correccion_negativos(catalog)
correccion_negativos(marketing)
print('negativos en orders')
revisar_negativos(orders)
print('negativos catalog')
revisar_negativos(catalog)
print('negativos Marketing')
revisar_negativos(marketing)


negativos en orders
cantidad:0
precio_unitario:0
monto_descuento:12551
monto_total:0

negativos catalog
costo_unitario:0

negativos Marketing
gasto:0



In [100]:
orders[orders['monto_total']<=0]

,id_pedido,id_usuario,fecha_hora_pedido,pais,dispositivo,fuente_referencia,nombre_producto,categoria_producto,cantidad,precio_unitario,monto_descuento,monto_total


**Nota:** Se realizan calculos comparativos dentro de un metodo con un ciclo for para detectar numeros negativos en columnas donde no deban estar estos. Se corrige usando el metodo .abs() para convertir todos los numeros negativos de los registros que presentaban este conflicto.  

en la revision se puede apreciar que ***monto_descuento*** aun conserva valores en 0 pero es normal ya que en esta columna es posible no tener un monto de descuento. 

In [101]:
orders.nunique()

id_pedido             25000
id_usuario             7642
fecha_hora_pedido       181
pais                      6
dispositivo               2
fuente_referencia         3
nombre_producto           7
categoria_producto        3
cantidad                  4
precio_unitario       19543
monto_descuento           4
monto_total           21536
dtype: int64

In [102]:
def valores_unicos(datos):
    columnas_obj = datos.select_dtypes(include=['object']).columns
    for columnas in columnas_obj:
         print(f'valores unicos {columnas} ')
         print(datos[columnas].unique())
         print()

valores_unicos(orders)

valores unicos id_pedido 
['order_0' 'order_1' 'order_2' ... 'order_24997' 'order_24998'
 'order_24999']

valores unicos id_usuario 
['user_6993' 'user_1329' 'user_3194' ... 'user_7965' 'user_76' 'user_670']

valores unicos pais 
['Argentina' 'Mexico' 'Colombia' 'mexico' 'colombia' 'argentina' nan]

valores unicos dispositivo 
['desktop' 'mobile' nan]

valores unicos fuente_referencia 
['organic' 'paid_search' 'social' nan]

valores unicos nombre_producto 
['Jacket-Winter-M' 'Tablet-Standard-64GB' 'Blender-XL-Red'
 'Laptop-Gaming-16GB' 'Sneakers-Urban-42' 'Phone-Pro-128GB'
 'Vacuum-Pro-Black' nan]

valores unicos categoria_producto 
['Moda' 'Electronica' 'Hogar' nan]



In [103]:
#valores unicos catalog
valores_unicos(catalog)

valores unicos nombre_producto 
['Laptop-Gaming-16GB' 'Phone-Pro-128GB' 'Tablet-Standard-64GB'
 'Blender-XL-Red' 'Vacuum-Pro-Black' 'Sneakers-Urban-42' 'Jacket-Winter-M']

valores unicos categoria_producto 
['Electrónica' 'Hogar' 'Moda']

valores unicos proveedor 
['Fuller, Pena and Myers' 'King Ltd' 'Bowers LLC' 'Long-Reid'
 'Rivera, Carr and Finley' 'Greene-Smith' 'Mcmillan-Rhodes']



In [104]:
#valores unicos marketing 
valores_unicos(marketing)


valores unicos pais 
['Mexico' 'Colombia' 'Argentina']

valores unicos id_campaña 
['organic_Mexico' 'paid_search_Mexico' 'social_Mexico' 'organic_Colombia'
 'paid_search_Colombia' 'social_Colombia' 'organic_Argentina'
 'paid_search_Argentina' 'social_Argentina']

valores unicos canal 
['organic' 'paid_search' 'social' nan]



In [105]:
#funcion para pasar todo a minusculas 
def normalizar_minusculas(datos):
    columnas_obj = datos.select_dtypes(include=['object']).columns
    for columnas in columnas_obj:
         print(f'Normalizando {columnas} ')
         datos[columnas] = datos[columnas].str.lower()
         print(datos[columnas].unique())
         print()
        
#normalizando tabla Orders        
normalizar_minusculas(orders)

Normalizando id_pedido 
['order_0' 'order_1' 'order_2' ... 'order_24997' 'order_24998'
 'order_24999']

Normalizando id_usuario 
['user_6993' 'user_1329' 'user_3194' ... 'user_7965' 'user_76' 'user_670']

Normalizando pais 
['argentina' 'mexico' 'colombia' nan]

Normalizando dispositivo 
['desktop' 'mobile' nan]

Normalizando fuente_referencia 
['organic' 'paid_search' 'social' nan]

Normalizando nombre_producto 
['jacket-winter-m' 'tablet-standard-64gb' 'blender-xl-red'
 'laptop-gaming-16gb' 'sneakers-urban-42' 'phone-pro-128gb'
 'vacuum-pro-black' nan]

Normalizando categoria_producto 
['moda' 'electronica' 'hogar' nan]



------------------------------------------------------------------------

In [106]:
#normalizando tabla catalog
catalog['categoria_producto'] = catalog['categoria_producto'].replace({'Electrónica': 'electronica'})
normalizar_minusculas(catalog)




Normalizando nombre_producto 
['laptop-gaming-16gb' 'phone-pro-128gb' 'tablet-standard-64gb'
 'blender-xl-red' 'vacuum-pro-black' 'sneakers-urban-42' 'jacket-winter-m']

Normalizando categoria_producto 
['electronica' 'hogar' 'moda']

Normalizando proveedor 
['fuller, pena and myers' 'king ltd' 'bowers llc' 'long-reid'
 'rivera, carr and finley' 'greene-smith' 'mcmillan-rhodes']



-------------------------------------------------------------

In [107]:
#normalizanod tabla marketing
normalizar_minusculas(marketing)

Normalizando pais 
['mexico' 'colombia' 'argentina']

Normalizando id_campaña 
['organic_mexico' 'paid_search_mexico' 'social_mexico' 'organic_colombia'
 'paid_search_colombia' 'social_colombia' 'organic_argentina'
 'paid_search_argentina' 'social_argentina']

Normalizando canal 
['organic' 'paid_search' 'social' nan]



**Nota:** se utiliza el metodo ***.unique()*** dentro de una función para poder explorar los datos y sus valores unicos e identificar si habia mas de una versión de lo que se ha registrado, se logra identificar que hay problemas en la tabla ***orders[Pais]*** en algunos registros (ejemplo: Mexico, mexico. Argentina, argentina, Colombia, colombia) se opta por pasar todos los recursos de las tablas a minusculas usando una funsion con el metodo ***str.lower()***. En esete mismo paso se encontaron registros con datos nulos o faltes de información, se procede a realizar una revision mas detallada con los nulos. 

Se realiza la sustitución de algunos registros de la tabla ***catalog['categoria_producto']*** ya que hay variación entre ellos ***electrónica*** de la misma tabla y ***electronica*** sin tilde de la tabla ***orders['categoria_producto']*** 

In [108]:
#conteo de duplicados orders
print(orders.duplicated().sum())
print(orders.duplicated(subset=['id_pedido']).sum())

100
100


In [109]:

#conteo de duplicados catalog
print(catalog.duplicated().sum())


0


In [110]:

#conteo de duplicados marketing
print(marketing.duplicated().sum())


0


In [111]:
orders = orders.drop_duplicates(subset=['id_pedido'], keep='first')
print(orders.duplicated().sum())

0


**Nota:** antes de pasar corregir los datos nulos se opta por primero detectar y eliminar datos duplicados para no trabajar con registros que de igual manera se tendran que eliminar. 

In [112]:

print(f'Cantidad de nulos en tabla orders')
print(orders.isnull().sum())




Cantidad de nulos en tabla orders
id_pedido               0
id_usuario              0
fecha_hora_pedido       0
pais                  300
dispositivo            20
fuente_referencia      30
nombre_producto        30
categoria_producto     80
cantidad               50
precio_unitario        50
monto_descuento        50
monto_total             0
dtype: int64


In [113]:
print(f'Cantidad de nulos en tabla Catalog')
print(catalog.isnull().sum())


Cantidad de nulos en tabla Catalog
nombre_producto       0
categoria_producto    0
costo_unitario        0
proveedor             0
dtype: int64


In [114]:
print(f'Cantidad de nulos en tabla Marketing')
print(marketing.isnull().sum())


Cantidad de nulos en tabla Marketing
fecha           0
pais            0
id_campaña      0
canal         101
gasto           0
dtype: int64


**Nota:** Se utilizan los metodos ***.isnull().sum()*** para detectar los nulos por columna de cada tabla. se llegan a detactar nulos en algunas columnas como lo son ***orders['pais','dispositivo','fuente_referencia','nombre_producto','cantidad','precio_unitario'.'monto_descuento']*** y ***marketing['canal']***. se procede a tomar medidas para eliminar o remplazar con valores logicos los nulos. 

In [115]:

#llenando nulos orders
orders['pais'].fillna('no_especidicado', inplace=True)
print(orders['pais'].isnull().sum())
print(orders['pais'].unique())

0
['argentina' 'mexico' 'colombia' 'no_especidicado']


In [116]:
orders['dispositivo'].fillna(orders['dispositivo'].mode()[0],inplace=True)
print(orders['dispositivo'].isnull().sum())
print(orders['dispositivo'].unique())

0
['desktop' 'mobile']


In [117]:
orders['fuente_referencia'].fillna('directo', inplace=True)
print(orders['fuente_referencia'].isnull().sum())
print(orders['fuente_referencia'].unique())

0
['organic' 'paid_search' 'social' 'directo']


In [118]:
orders['categoria_producto'].fillna('otros', inplace=True)
print(orders['categoria_producto'].isnull().sum())
print(orders['categoria_producto'].unique())

0
['moda' 'electronica' 'hogar' 'otros']


In [119]:
orders['nombre_producto'].fillna('no_especificado', inplace=True)
print(orders['nombre_producto'].isnull().sum())
print(orders['nombre_producto'].unique())

0
['jacket-winter-m' 'tablet-standard-64gb' 'blender-xl-red'
 'laptop-gaming-16gb' 'sneakers-urban-42' 'phone-pro-128gb'
 'vacuum-pro-black' 'no_especificado']


In [120]:
orders['monto_descuento'].fillna('0', inplace=True)
orders['monto_descuento'] = orders['monto_descuento'].astype('float64')
print((orders['monto_descuento'] < 0).sum())


0


In [121]:

ambos_nulos = (orders[orders['precio_unitario'].isnull() | orders['cantidad'].isnull()])
print(len(ambos_nulos))


50


In [122]:
ambos_nulos

,id_pedido,id_usuario,fecha_hora_pedido,pais,dispositivo,fuente_referencia,nombre_producto,categoria_producto,cantidad,precio_unitario,monto_descuento,monto_total
74,order_74,user_6172,2025-01-15,argentina,desktop,organic,sneakers-urban-42,otros,NaN,NaN,0.0,595.85
75,order_75,user_6588,2025-06-19,colombia,mobile,organic,sneakers-urban-42,otros,NaN,NaN,0.0,458.15
76,order_76,user_3193,2025-03-17,argentina,mobile,paid_search,laptop-gaming-16gb,otros,NaN,NaN,0.0,319.75
77,order_77,user_775,2025-06-24,argentina,desktop,social,vacuum-pro-black,otros,NaN,NaN,0.0,227.55
78,order_78,user_2702,2025-03-19,argentina,desktop,paid_search,phone-pro-128gb,otros,NaN,NaN,0.0,432.39
79,order_79,user_6438,2025-03-22,colombia,desktop,paid_search,sneakers-urban-42,otros,NaN,NaN,0.0,527.15
80,order_80,user_6790,2025-01-03,mexico,desktop,organic,vacuum-pro-black,otros,NaN,NaN,0.0,711.18
81,order_81,user_232,2025-05-16,mexico,mobile,paid_search,tablet-standard-64gb,otros,NaN,NaN,0.0,41.08
82,order_82,user_5665,2025-02-07,colombia,mobile,paid_search,sneakers-urban-42,otros,NaN,NaN,0.0,431.55
83,order_83,user_5225,2025-04-25,argentina,mobile,organic,sneakers-urban-42,otros,NaN,NaN,0.0,279.35


In [123]:
# Calcular precio promedio por producto de los registros válidos
precio_promedio_producto = orders[orders['precio_unitario'].notnull()].groupby('nombre_producto')['precio_unitario'].mean()

print("Precios promedio por producto:")
print(precio_promedio_producto)


Precios promedio por producto:
nombre_producto
blender-xl-red          259.157926
jacket-winter-m         259.818113
laptop-gaming-16gb      261.359960
no_especificado         264.574000
phone-pro-128gb         259.543365
sneakers-urban-42       257.232930
tablet-standard-64gb    257.533126
vacuum-pro-black        261.018463
Name: precio_unitario, dtype: float64


In [124]:
# Llenar precio_unitario con el promedio del producto
orders['precio_unitario'] = orders['precio_unitario'].fillna(orders['nombre_producto'].map(precio_promedio_producto))

# Calcular cantidad usando: cantidad = (monto_total + monto_descuento) / precio_unitario
mask_cantidad_nula = orders['cantidad'].isnull()
orders.loc[mask_cantidad_nula, 'cantidad'] = (orders.loc[mask_cantidad_nula, 'monto_total'] + orders.loc[mask_cantidad_nula, 'monto_descuento']) / orders.loc[mask_cantidad_nula, 'precio_unitario']
orders['cantidad'] = orders['cantidad'].round().astype('int64')
# Verificar que no quedan nulos
print("Nulos restantes:")
print(orders[['precio_unitario', 'cantidad']].isnull().sum())

Nulos restantes:
precio_unitario    0
cantidad           0
dtype: int64


In [125]:
ambos_nulos = (orders[orders['precio_unitario'].isnull() | orders['cantidad'].isnull()])
ambos_nulos


,id_pedido,id_usuario,fecha_hora_pedido,pais,dispositivo,fuente_referencia,nombre_producto,categoria_producto,cantidad,precio_unitario,monto_descuento,monto_total


In [126]:
orders['cantidad'].head(10) 

0    2
1    1
2    2
3    1
4    1
5    1
6    2
7    2
8    2
9    1
Name: cantidad, dtype: int64

In [127]:
print(f'Cantidad de nulos en tabla orders')
print(orders.isnull().sum())


Cantidad de nulos en tabla orders
id_pedido             0
id_usuario            0
fecha_hora_pedido     0
pais                  0
dispositivo           0
fuente_referencia     0
nombre_producto       0
categoria_producto    0
cantidad              0
precio_unitario       0
monto_descuento       0
monto_total           0
dtype: int64


In [128]:
marketing['canal'].fillna('directo', inplace=True)
print(marketing['canal'].isnull().sum())
print(marketing['canal'].unique())

0
['organic' 'paid_search' 'social' 'directo']


In [129]:
catalog.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7 entries, 0 to 6
Data columns (total 4 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   nombre_producto     7 non-null      object 
 1   categoria_producto  7 non-null      object 
 2   costo_unitario      7 non-null      float64
 3   proveedor           7 non-null      object 
dtypes: float64(1), object(3)
memory usage: 352.0+ bytes


In [130]:

# PASO CRÍTICO: Eliminar la columna 'categoria_producto' de 'orders' antes de unir
# Esto evita que se creen las columnas '_x' y '_y'
if 'categoria_producto' in orders.columns:
    orders = orders.drop(columns=['categoria_producto'])
    print("✅ Columna original eliminada para evitar duplicados.")

# Paso 2: Realizar el Merge (Unión)
# Ahora solo habrá una columna 'categoria_producto' (la del catálogo)
orders = orders.merge(
    catalog[['nombre_producto', 'categoria_producto']], 
    on='nombre_producto', 
    how='left'
)

# Paso 3: Manejar valores faltantes
orders['categoria_producto'] = orders['categoria_producto'].fillna('otros')

# Paso 4: Verificar
print(f"\nColumnas actuales en 'orders': {orders.columns.tolist()}")
print("\nDistribución de categorías:")
print(orders['categoria_producto'].value_counts())



✅ Columna original eliminada para evitar duplicados.

Columnas actuales en 'orders': ['id_pedido', 'id_usuario', 'fecha_hora_pedido', 'pais', 'dispositivo', 'fuente_referencia', 'nombre_producto', 'cantidad', 'precio_unitario', 'monto_descuento', 'monto_total', 'categoria_producto']

Distribución de categorías:
hogar          8355
moda           8324
electronica    8291
otros            30
Name: categoria_producto, dtype: int64


In [131]:
orders.head()

,id_pedido,id_usuario,fecha_hora_pedido,pais,dispositivo,fuente_referencia,nombre_producto,cantidad,precio_unitario,monto_descuento,monto_total,categoria_producto
0,order_0,user_6993,2025-05-22,argentina,desktop,organic,jacket-winter-m,2,332.69,0.0,665.37,moda
1,order_1,user_1329,2025-06-15,mexico,desktop,paid_search,tablet-standard-64gb,1,176.86,5.0,171.86,electronica
2,order_2,user_3194,2025-05-02,argentina,desktop,social,blender-xl-red,2,102.99,10.0,195.99,hogar
3,order_3,user_4510,2025-06-09,colombia,mobile,social,tablet-standard-64gb,1,257.87,15.0,242.87,electronica
4,order_4,user_5044,2025-03-30,argentina,desktop,paid_search,blender-xl-red,1,336.28,0.0,336.28,hogar


**Nota:** se opta por agregar etiquetas a los nulos para lograr identificarlos sin perder valores importantes a la hora de analizar y representar en los dashboards (*ejemplo: directo, no_especificado y otros*).En el caso especifico de **dispositivo** se opto por imputar la media de los registros de dicha columna. En los casos de columnas numericas se opto por sacar los valores promedios de cada articulo vendido que contava con valores nulos y asi se agrego el promedio como precio_unitario y de esa manera poder calcular la cantidad de produdctos adquiridos en la compra diviendo el monto total entre el precio unitario, despues de esto se realiza una conversión para no dejar cantidades con decimales y dejar el tipo en numeros enteros como ***int64*** y como ultimo se tomo la decision de dejar los nulos en descuentos como 0. 

Se tedectan anomalias entre la realación de las columnas ***nombre_producto*** y ***categoria_producto*** de tabla orders ya que algunos productos que si cuentan con categoria se encontraban con registgro anterior mentue como nulo y despues de normalizar paso con la categoria ***otros***, se opta por eliminar la columna original para evitar columnas duplicadas y se procede hacer un ***.merge()*** para relacionar correctamente las columnas ***nombre_produto*** de ***orders*** con ***categoria_producto*** de ***catalog***

In [132]:
def variables_categoricas_count(datos):
    columnas_obj = datos.select_dtypes(include=['object']).columns
    columnas_a_ignorar = ['id_usuario','id_pedido']
    for columnas in columnas_obj:
        if columnas not in columnas_a_ignorar:
             print(f'conteo de registros {columnas} ')
             print(datos[columnas].value_counts())
             print()

variables_categoricas_count(orders)
print ("revision Marketing")
print ()
variables_categoricas_count(marketing)

conteo de registros pais 
mexico             8341
colombia           8304
argentina          8055
no_especidicado     300
Name: pais, dtype: int64

conteo de registros dispositivo 
desktop    12731
mobile     12269
Name: dispositivo, dtype: int64

conteo de registros fuente_referencia 
social         8402
organic        8288
paid_search    8280
directo          30
Name: fuente_referencia, dtype: int64

conteo de registros nombre_producto 
blender-xl-red          4179
jacket-winter-m         4176
vacuum-pro-black        4176
sneakers-urban-42       4148
laptop-gaming-16gb      2782
tablet-standard-64gb    2769
phone-pro-128gb         2740
no_especificado           30
Name: nombre_producto, dtype: int64

conteo de registros categoria_producto 
hogar          8355
moda           8324
electronica    8291
otros            30
Name: categoria_producto, dtype: int64

revision Marketing

conteo de registros pais 
argentina    540
colombia     540
mexico       540
Name: pais, dtype: int64

conte

In [133]:
def variables_categoricas_descr(datos):
    columnas_obj = datos.select_dtypes(include=['object']).columns
    columnas_a_ignorar = ['id_pedido']
    for columnas in columnas_obj:
        if columnas not in columnas_a_ignorar:
             print(f'Datos de {columnas} ')
             print(datos[columnas].describe())
             print()

variables_categoricas_descr(orders)
print ("Describe Marketing")
print ()
variables_categoricas_descr(marketing)

Datos de id_usuario 
count         25000
unique         7642
top       user_7769
freq             11
Name: id_usuario, dtype: object

Datos de pais 
count      25000
unique         4
top       mexico
freq        8341
Name: pais, dtype: object

Datos de dispositivo 
count       25000
unique          2
top       desktop
freq        12731
Name: dispositivo, dtype: object

Datos de fuente_referencia 
count      25000
unique         4
top       social
freq        8402
Name: fuente_referencia, dtype: object

Datos de nombre_producto 
count              25000
unique                 8
top       blender-xl-red
freq                4179
Name: nombre_producto, dtype: object

Datos de categoria_producto 
count     25000
unique        4
top       hogar
freq       8355
Name: categoria_producto, dtype: object

Describe Marketing

Datos de pais 
count          1620
unique            3
top       argentina
freq            540
Name: pais, dtype: object

Datos de id_campaña 
count                 1620
uniq

**nota:** una vez corregidos los datos se utilizan los metodos ***value_counts()*** y ***describe()*** para dar una revision general de los datos categoricos que cuentan las tablas. 

In [134]:

     orders['mes'] = pd.to_datetime(orders['fecha_hora_pedido']).dt.to_period('M').astype(str)
     print(orders['mes'].value_counts())


2025-01    4326
2025-03    4269
2025-05    4254
2025-04    4222
2025-06    4176
2025-02    3753
Name: mes, dtype: int64


In [135]:
temp = orders.merge(
    catalog[['nombre_producto', 'costo_unitario']],
    on='nombre_producto',
    how='left'
)

orders['es_anomalia'] = ((temp['cantidad']>100)| (temp['precio_unitario'] < temp['costo_unitario'].fillna(0)))
print(orders['es_anomalia'].value_counts())
print()
orders['tipo_anomalia']= 'normal'
orders.loc[temp['cantidad']>100, 'tipo_anomalia'] = 'cantidad_masiva'
orders.loc[
    (temp['cantidad']<=100) & (temp['precio_unitario']<temp['costo_unitario'].fillna(0)),'tipo_anomalia'
] = 'precio_bajo_costo'

print(orders['tipo_anomalia'].value_counts())

False    20608
True      4392
Name: es_anomalia, dtype: int64

normal               20608
precio_bajo_costo     4382
cantidad_masiva         10
Name: tipo_anomalia, dtype: int64


In [136]:
orders.head()

,id_pedido,id_usuario,fecha_hora_pedido,pais,dispositivo,fuente_referencia,nombre_producto,cantidad,precio_unitario,monto_descuento,monto_total,categoria_producto,mes,es_anomalia,tipo_anomalia
0,order_0,user_6993,2025-05-22,argentina,desktop,organic,jacket-winter-m,2,332.69,0.0,665.37,moda,2025-05,False,normal
1,order_1,user_1329,2025-06-15,mexico,desktop,paid_search,tablet-standard-64gb,1,176.86,5.0,171.86,electronica,2025-06,False,normal
2,order_2,user_3194,2025-05-02,argentina,desktop,social,blender-xl-red,2,102.99,10.0,195.99,hogar,2025-05,True,precio_bajo_costo
3,order_3,user_4510,2025-06-09,colombia,mobile,social,tablet-standard-64gb,1,257.87,15.0,242.87,electronica,2025-06,False,normal
4,order_4,user_5044,2025-03-30,argentina,desktop,paid_search,blender-xl-red,1,336.28,0.0,336.28,hogar,2025-03,False,normal


**Nota:** Se decide separar de la columna ***orders['fecha_hora_pedido']*** el ***mes*** en una columna nueva para poder manipular con mayor

In [137]:
# Extraer mes
marketing['mes'] = pd.to_datetime(marketing['fecha']).dt.to_period('M').astype(str)

# Agrupar por mes+pais — elimina los 4 canales, queda 1 fila por combinación
marketing_mes_clean = marketing.groupby(['mes','pais'])['gasto'].sum().reset_index()

marketing_mes_clean.to_csv('marketing_mes_clean.csv', index=False)
print('marketing_mes_clean.csv listo —', marketing_mes_clean.shape[0], 'filas,', marketing_mes_clean.shape[1], 'columnas')

# ── 3. VERIFICAR ──────────────────────────────────────────────────────
print()
print('Gasto original:  $', round(marketing['gasto'].sum(), 2))
print('Gasto agrupado:  $', round(marketing_mes_clean['gasto'].sum(), 2))

marketing_mes_clean.csv listo — 18 filas, 3 columnas

Gasto original:  $ 2871843.53
Gasto agrupado:  $ 2871843.53


---
**📦 Exportación**: Una vez finalizada la limpieza, se exportan los datasets para utilizarlos en la última etapa del proyecto.

**Nota:** Se detectan anomalias en cuanto a precios y cantidades, se requiere solicitar la confirmación de los datos a los departamentos correspondientes ya que los valores detectados pueden deberse a promciones que se dieron en el momento o compras masivas coporporativas. Algunos de los valores detectados son ***laptop-gaming-16gb*** con cantidades de **10,000** y **20,000** unidades donde en algunos casos ***precio_unitario*** es menor al ***costo_unitario*** ( ejemplo: order_3521 | cantidad 10,000 | precio_unitario 43.14 | costo_unitario 280.68 ) esto ultimo generando profiits negativos enormes  que alteran la distribucion de datos en toda la categoria ***electronica***

Se opta por separar los datos en catergorias adicionales (***es_anomalia*** o si la compra es ***normal***) esto con el fina de no perder los datos y graficar ambos datos, dejando de lado las posibles "anomalias" detectadas y otro con todos los datos en general. Tambien se opta por agregar una columna separando la fecha por meses para poder tener una mayor control de los datos de manera mensual tanto en ***marketing*** como en ***orders***

In [138]:
# exportar datasets
orders.to_csv('orders_clean.csv', index=False)
catalog.to_csv('catalog_clean.csv', index=False)
marketing.to_csv('marketing_clean.csv', index=False)

**Nota:** se genera al igual que los archivos csv clean como adicional ***marketing_mes_clean.csv*** esto con el fina de poder manipular con mas rapides los datos en **Tableau**

---

## 🔹 Paso 2: Analizar si el negocio es rentable

### 2.1 Cálculo de KPIs principales

**🎯 Objetivo:** Calcular los indicadores clave del negocio para evaluar ingresos, costos y rentabilidad.

Se usan los 3 datasets (`orders`, `catalog`, `marketing_spend`):

**📊 Parte 1: Rentabilidad del negocio**
- ¿Cuál es el ingreso total (revenue)? 
- ¿Cuál es el costo total? 
- ¿Cuánto se ha invertido en marketing? 
- ¿El negocio es rentable? (calcular profit)  

---

**📈 Parte 2: Comportamiento de ventas**
- ¿Cuál es el ticket promedio por orden? 
- ¿Cuál es la cantidad promedio de productos por orden? 
- ¿Cuál es el producto más vendido?
- ¿Cuánto se ha gastado en marketing por canal? 

In [139]:
# tu código aquí
Ingreso_total = orders['monto_total'].sum()
print(f'Ingrso total: ${Ingreso_total:.2f}')

Ingrso total: $51989756.86


In [140]:
Descuento_total = orders['monto_descuento'].sum()
Costo_total_marketing = (marketing['gasto'].sum())
costos_referencia = catalog.set_index('nombre_producto')['costo_unitario']
costo_total_producto = (orders['nombre_producto'].map(costos_referencia) * orders['cantidad']).sum()
Costo_total = Costo_total_marketing + costo_total_producto
print(f'Costo total: ${Costo_total:.2f}')

Costo total: $46003121.12


In [141]:
print(f'Inversion en Marketing: ${Costo_total_marketing:.2f}')

Inversion en Marketing: $2871843.53


In [142]:
Ingresos_sin_descuento = Ingreso_total - Descuento_total
print(f'Ingreso total sin descuento: ${Ingresos_sin_descuento:.2f}')

Ingreso total sin descuento: $51877416.86


In [143]:
Profit_Neto = Ingreso_total - Costo_total
print(f'Profit Neto: ${Profit_Neto:.2f}')

Profit Neto: $5986635.74


In [144]:
Profit_Bruto = Ingreso_total - costo_total_producto
print(f'Profit Bruto: ${Profit_Bruto:.2f}')

Profit Bruto: $8858479.27


In [145]:
#calculando ticket promedio
ticket_promedio = orders['monto_total'].mean()
print(f'Ticket promedio: ${ticket_promedio:.2f}')

Ticket promedio: $2079.59


In [146]:
#cantidad promedio por orden 
cantidad_promedio = orders['cantidad'].mean()
print(f'Cantidad promedio por orden: {cantidad_promedio:.0f}')

Cantidad promedio por orden: 7


In [147]:

#producto mas vendido
producto_mas_vendido = orders.groupby('nombre_producto')['cantidad'].sum().idxmax()
cantidad_mas_vendida = orders.groupby('nombre_producto')['cantidad'].sum().max()
print(f'producto mas vendido: {producto_mas_vendido}')
print(f'unidades vendidas: {cantidad_mas_vendida:.0f}')


producto mas vendido: laptop-gaming-16gb
unidades vendidas: 144205


In [148]:
gasto_marketing_canal = marketing.groupby('canal')['gasto'].sum()
print(f'Gasto total por canal en Marketing')
print(gasto_marketing_canal)

Gasto total por canal en Marketing
canal
directo        177179.10
organic        913533.01
paid_search    863088.21
social         918043.21
Name: gasto, dtype: float64


**Nota:** de acuerdo a los valores  arrojados en los calculos podemos apreciar un profit positivo con una cantidad muy considerablemente favorable de $5874101.84, con esto podemos concluir que el negocio es rentable. 
Los calculos realizados anterior mente ya pueden ser usados como indicadores que puedan ayudar a comprender mejor el comportamiento del negocio. 


---

## 🔹 Paso 3: Entender dónde se pierden los usuarios (funnel de conversión)

**🎯 Objetivo:** Analizar el comportamiento de los usuarios para identificar en qué etapa del proceso se pierden.


⚙️**Conexión a la base de datos**:  
Se ejecuta la línea de configuración para conectar con la base de datos y aplicar consultas SQL en la tabla **events**.

---

**📊 Parte 1: Construcción del funnel**
- ¿Cuántos usuarios llegan a cada etapa del funnel?  
- Se calcula el número de usuarios únicos por `nombre_evento`  
- Se ordenan los eventos según el flujo del usuario  

---

**📉 Parte 2: Análisis de conversión**
- Se calcula la tasa de conversión entre cada paso del funnel  
- Se identifica en qué etapa se pierde la mayor cantidad de usuarios  
- ¿Cuál es la tasa de conversión final?
---

In [149]:
import pandas as pd
from sqlalchemy import create_engine

# ======================
# Conexión (NO modificar)
# ======================
db_config = {
    'user': 'practicum_student',
    'pwd': 'QnmDH8Sc2TQLvy2G3Vvh7',
    'host': 'yp-trainers-practicum.cluster-czs0gxyx2d8w.us-east-1.rds.amazonaws.com',
    'port': 5432,
    'db': 'data-analyst-production-db-en'
}

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(
    db_config['user'],
    db_config['pwd'],
    db_config['host'],
    db_config['port'],
    db_config['db']
)

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

In [150]:
# Explorar tabla events
# =========================
query_events = '''
SELECT *
FROM events;
'''
events = pd.read_sql(query_events, con=engine)
events.head()

,id_usuario,id_sesion,nombre_evento,timestamp_evento,pais,dispositivo,fuente_referencia,categoria_producto
0,user_6772,6a97f2af-32ae-4186-8c92-04025be1a27b,first_visit,2025-05-17,Colombia,desktop,organic,Moda
1,user_5883,369b767c-1c33-4b2f-a652-c7c0ef92cfc9,add_to_cart,2025-02-23,Mexico,mobile,social,Hogar
2,user_5946,60039041-e78b-474c-87b3-c0b7e9c30708,add_payment_info,2025-05-15,Colombia,desktop,social,Electronica
3,user_827,18252a64-f389-4ef7-9e58-dadad4a3491e,purchase,2025-03-31,Mexico,mobile,social,Moda
4,user_2361,221b364e-cdc5-4668-b698-18d5ba849a67,first_visit,2025-01-22,Argentina,desktop,paid_search,Electronica


In [151]:
# PARTE 1: Totales del funnel
# ======================

query_totals = '''
SELECT 
    nombre_evento,
    COUNT(DISTINCT id_usuario) as usuarios_unicos
FROM events
GROUP BY nombre_evento
ORDER BY usuarios_unicos DESC;
'''

totals = pd.read_sql(query_totals, con=engine)
totals

,nombre_evento,usuarios_unicos
0,first_visit,7796
1,add_to_cart,7634
2,select_item,7582
3,begin_checkout,7208
4,add_payment_info,6250
5,purchase,6240


**Nota:** Se realiza un conteo unico por usuarios que se encuentren agrupados por la columna ***nombre_evento*** esto para ver el volumen general de cada etapa.

In [152]:
# PARTE 2: Conversiones
# ======================

query_conversion = '''

SELECT 

    -- Tasas de conversión
    ROUND(
        COUNT(DISTINCT CASE WHEN nombre_evento = 'add_to_cart' THEN id_usuario END) * 100.0 / 
        COUNT(DISTINCT CASE WHEN nombre_evento = 'first_visit' THEN id_usuario END), 2
    ) AS conversion_add_to_cart,
    
    ROUND(
        COUNT(DISTINCT CASE WHEN nombre_evento = 'select_item' THEN id_usuario END) * 100.0 / 
        COUNT(DISTINCT CASE WHEN nombre_evento = 'add_to_cart' THEN id_usuario END), 2
    ) AS conversion_select_item,

    ROUND(
        COUNT(DISTINCT CASE WHEN nombre_evento = 'begin_checkout' THEN id_usuario END) * 100.0 / 
        COUNT(DISTINCT CASE WHEN nombre_evento = 'select_item' THEN id_usuario END), 2
    ) AS conversion_begin_check_out,

    ROUND(
        COUNT(DISTINCT CASE WHEN nombre_evento = 'add_payment_info' THEN id_usuario END) * 100.0 / 
        COUNT(DISTINCT CASE WHEN nombre_evento = 'begin_checkout' THEN id_usuario END), 2 
    ) AS conversion_add_payment_info,

    ROUND(
        COUNT(DISTINCT CASE WHEN nombre_evento = 'purchase' THEN id_usuario END) * 100.0 / 
        COUNT(DISTINCT CASE WHEN nombre_evento = 'add_payment_info' THEN id_usuario END), 2
    ) AS conversion_purchase,

    ROUND(
        COUNT(DISTINCT CASE WHEN nombre_evento = 'purchase' THEN id_usuario END) * 100 / 
        COUNT(DISTINCT CASE WHEN nombre_evento = 'first_visit' THEN id_usuario END), 2
    ) AS Conversion_Final
    
FROM events
WHERE nombre_evento IN ('first_visit', 'add_to_cart', 'select_item', 'begin_checkout','add_payment_info', 'purchase');


'''

conversion = pd.read_sql(query_conversion, con=engine)
conversion

,conversion_add_to_cart,conversion_select_item,conversion_begin_check_out,conversion_add_payment_info,conversion_purchase,conversion_final
0,97.92,99.32,95.07,86.71,99.84,80.0


**NOTA:** Se logra apreción que la mayor perdida de usuarios se encuentran en la etapa de ***add_payment_info*** donde solo el ***87.71%*** de los usuarios que habían llegado a al a etapa de ***begin_check_out*** lograron continuar con camino hasta llegar a ***purchase***, se perdio el ***13.29%*** solo entre esas etapas. 

---

## 🔹 Paso 4: Evaluar si los usuarios regresan (retención por cohortes)

**🎯 Objetivo:** Analizar la retención de usuarios para entender si regresan después de registrarse.

**Tablas**

- `users` 
- `user_activity` 

---
1. Se identifica la cohorte de cada usuario según el **mes de registro**.


2. Se calcula la retención semanal: cuántos usuarios **se mantienen activos** en cada semana desde su registro.
   - `retenido_w1`: usuarios activos en la semana 1  
   - `retenido_w2`: usuarios activos en la semana 2  
   - `retenido_w3`: usuarios activos en la semana 3  


3. Se calcula el porcentaje de retención para cada semana, dividiendo los usuarios retenidos entre los clientes iniciales de la cohorte:  
   - `semana_1`: porcentaje de usuarios retenidos en la semana 1  
   - `semana_2`: porcentaje de usuarios retenidos en la semana 2  
   - `semana_3`: porcentaje de usuarios retenidos en la semana 3  

Se revisa que la columna de fecha esté en formato correcto (`DATE`).  
Se realiza la conversión usando: `CAST(fecha_registro AS DATE)`

In [153]:
# Explorar tabla users
# =========================
query_users = '''
SELECT *
FROM users;
'''
users = pd.read_sql(query_users, con=engine)
users.head(3)

,id_usuario,fecha_registro,país,dispositivo,tipo_plan
0,user_0,2025-01-29,Mexico,mobile,free
1,user_1,2025-01-07,Mexico,mobile,free
2,user_2,2025-03-12,Argentina,mobile,free


In [154]:
# Explorar tabla user_activity
# =========================
query_user_activity = '''
SELECT *
FROM user_activity;
'''
user_activity = pd.read_sql(query_user_activity, con=engine)
user_activity.head(10)

,id_usuario,fecha_actividad,dias_despues_registro,activo
0,user_0,2025-02-05,7,0
1,user_0,2025-02-12,14,1
2,user_0,2025-02-19,21,1
3,user_0,2025-02-26,28,0
4,user_1,2025-01-14,7,0
5,user_1,2025-01-21,14,0
6,user_1,2025-01-28,21,1
7,user_1,2025-02-04,28,0
8,user_2,2025-03-19,7,0
9,user_2,2025-03-26,14,1


In [155]:
# Retención por cohortes
# ======================

query_cohort_retention_final = '''
WITH cohortes AS (
    SELECT 
        id_usuario,
        DATE_TRUNC('month', CAST(fecha_registro AS DATE)) as mes_cohorte,
        fecha_registro
    FROM users
),
usuarios_por_cohorte AS (
    SELECT 
        mes_cohorte,
        COUNT(DISTINCT id_usuario) as usuarios_iniciales
    FROM cohortes
    GROUP BY mes_cohorte
),
actividad_semanal AS (
    SELECT 
        c.id_usuario,
        c.mes_cohorte,
        ua.dias_despues_registro,
        CASE 
            WHEN ua.dias_despues_registro BETWEEN 1 AND 7 THEN 'semana_1'
            WHEN ua.dias_despues_registro BETWEEN 8 AND 14 THEN 'semana_2'
            WHEN ua.dias_despues_registro BETWEEN 15 AND 21 THEN 'semana_3'
            WHEN ua.dias_despues_registro BETWEEN 22 AND 28 THEN 'semana_4'
        END as periodo,
        COALESCE(ua.activo, 0) as activo
    FROM cohortes c
    LEFT JOIN user_activity ua ON c.id_usuario = ua.id_usuario
    WHERE ua.dias_despues_registro <= 28
)
SELECT 
    DATE(upc.mes_cohorte),
    upc.usuarios_iniciales AS Usuarios_iniciales,
    COUNT(DISTINCT CASE WHEN periodo = 'semana_1' AND activo = 1 THEN id_usuario END) as retenido_w1,
    COUNT(DISTINCT CASE WHEN periodo = 'semana_2' AND activo = 1 THEN id_usuario END) as retenido_w2,
    COUNT(DISTINCT CASE WHEN periodo = 'semana_3' AND activo = 1 THEN id_usuario END) as retenido_w3,
    COUNT(DISTINCT CASE WHEN periodo = 'semana_4' AND activo = 1 THEN id_usuario END) as retenido_w4
FROM usuarios_por_cohorte upc 
LEFT JOIN actividad_semanal as_ ON upc.mes_cohorte = as_.mes_cohorte
GROUP BY upc.mes_cohorte, upc.usuarios_iniciales
ORDER BY upc.mes_cohorte;
'''

# Ejecutar la consulta
cohorte_final = pd.read_sql(query_cohort_retention_final, con=engine)
cohorte_final

,date,usuarios_iniciales,retenido_w1,retenido_w2,retenido_w3,retenido_w4
0,2025-01-01,1627,697,668,656,671
1,2025-02-01,1444,611,609,635,575
2,2025-03-01,1636,677,705,690,673
3,2025-04-01,1606,680,697,663,652
4,2025-05-01,1687,695,676,706,679


In [156]:
# Retención por cohortes
# ======================

query_cohort_retention_final = '''
WITH cohortes AS (
    SELECT 
        id_usuario,
        DATE_TRUNC('month', CAST(fecha_registro AS DATE)) as mes_cohorte,
        fecha_registro
    FROM users
),
usuarios_por_cohorte AS (
    SELECT 
        mes_cohorte,
        COUNT(DISTINCT id_usuario) as usuarios_iniciales
    FROM cohortes
    GROUP BY mes_cohorte
),
actividad_semanal AS (
    SELECT 
        c.id_usuario,
        c.mes_cohorte,
        ua.dias_despues_registro,
        CASE 
            WHEN ua.dias_despues_registro BETWEEN 1 AND 7 THEN 'semana_1'
            WHEN ua.dias_despues_registro BETWEEN 8 AND 14 THEN 'semana_2'
            WHEN ua.dias_despues_registro BETWEEN 15 AND 21 THEN 'semana_3'
            WHEN ua.dias_despues_registro BETWEEN 22 AND 28 THEN 'semana_4'
        END as periodo,
        COALESCE(ua.activo, 0) as activo
    FROM cohortes c
    LEFT JOIN user_activity ua ON c.id_usuario = ua.id_usuario
    WHERE ua.dias_despues_registro <= 28
)
SELECT 
    DATE(upc.mes_cohorte),
    upc.usuarios_iniciales*100/upc.usuarios_iniciales AS Usuarios_iniciales,
    ((COUNT(DISTINCT CASE WHEN periodo = 'semana_1' AND activo = 1 THEN id_usuario END))*100)/ upc.usuarios_iniciales as retenido_w1,
    ((COUNT(DISTINCT CASE WHEN periodo = 'semana_2' AND activo = 1 THEN id_usuario END))*100)/ upc.usuarios_iniciales as retenido_w2,
    ((COUNT(DISTINCT CASE WHEN periodo = 'semana_3' AND activo = 1 THEN id_usuario END))*100)/ upc.usuarios_iniciales as retenido_w3,
    ((COUNT(DISTINCT CASE WHEN periodo = 'semana_4' AND activo = 1 THEN id_usuario END))*100)/ upc.usuarios_iniciales as retenido_w4
FROM usuarios_por_cohorte upc 
LEFT JOIN actividad_semanal as_ ON upc.mes_cohorte = as_.mes_cohorte
GROUP BY upc.mes_cohorte, upc.usuarios_iniciales
ORDER BY upc.mes_cohorte;
'''

# Ejecutar la consulta
cohorte_final = pd.read_sql(query_cohort_retention_final, con=engine)
cohorte_final

,date,usuarios_iniciales,retenido_w1,retenido_w2,retenido_w3,retenido_w4
0,2025-01-01,100,42,41,40,41
1,2025-02-01,100,42,42,43,39
2,2025-03-01,100,41,43,42,41
3,2025-04-01,100,42,43,41,40
4,2025-05-01,100,41,40,41,40


**Nota:** Se puede apreciar que hay una retención consistente del aproxicamente 40-43% de usuarios activos cada semana, las cohortes presentan estabilidad ya que los porsentajes que se muestran son muy similares entre los diferentes meses evaluados. 

---

## 🔹 Paso 5: Validar si los cambios generan impacto (test estadístico)

🎯 **Objetivo:** Evaluar si la modificación en la UI del checkout impacta la **tasa de conversión de compra**.

---

1. **Analizar el dataset** `experiment_checkout_ui.csv` para identificar la métrica principal **conversion**.
   - La métrica **conversion** es 1 si el usuario completó la compra, 0 si no.    
2. **Plantear la hipótesis estadística**     
3. **Aplicar el test estadístico adecuado** 
4. **Interpretar el resultado**  

---
Hipótesis estadística
   - **H₀ (Hipótesis nula):** Los cambios en el UI del checkout NO impactan en la tasa de conversión 
   - **H₁ (Hipótesis alternativa):**  Los cambios en el UI del checkout SÍ generan una diferencia significativa
   
**Test estadístico:** Z-Test  
**Nivel de significancia alpha:**  α = 0.05

In [157]:
# tu código aquí
experiment = pd.read_csv('https://practicum-content.s3.amazonaws.com/datasets/experiment_checkout_ui.csv')


In [158]:
experiment.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 7 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   id_usuario       10000 non-null  object 
 1   variante         10000 non-null  object 
 2   convirtio        10000 non-null  int64  
 3   dispositivo      10000 non-null  object 
 4   pais             10000 non-null  object 
 5   duracion_sesion  10000 non-null  float64
 6   timestamp        10000 non-null  object 
dtypes: float64(1), int64(1), object(5)
memory usage: 547.0+ KB


In [159]:
experiment.head(3)

,id_usuario,variante,convirtio,dispositivo,pais,duracion_sesion,timestamp
0,exp_user_0,tratamiento,0,mobile,Argentina,114.41,2025-03-28
1,exp_user_1,tratamiento,0,desktop,Mexico,170.03,2025-01-15
2,exp_user_2,control,1,mobile,Colombia,140.21,2025-03-18


In [160]:
experiment['timestamp'] = pd.to_datetime(experiment['timestamp'],errors='coerce')
print(experiment['timestamp'].dtype)

datetime64[ns]


In [161]:
experiment.duplicated(subset='id_usuario').sum()

0

In [162]:
valores_unicos(experiment)

valores unicos id_usuario 
['exp_user_0' 'exp_user_1' 'exp_user_2' ... 'exp_user_9997'
 'exp_user_9998' 'exp_user_9999']

valores unicos variante 
['tratamiento' 'control']

valores unicos dispositivo 
['mobile' 'desktop']

valores unicos pais 
['Argentina' 'Mexico' 'Colombia']



In [163]:
experiment['convirtio'].unique()

array([0, 1])

**Nota:** Antes de pasar a las pruebas estadisticas primero se opta por realizar una exploración de datos, para poder determinar si convertir datos, limpiar y corregir lo que sea necesario para hacer una prubea confiable.

In [164]:
conversiones = experiment.groupby('variante')['convirtio'].sum()
totales = experiment.groupby('variante')['convirtio'].count()
print(f'total de conversiones por {conversiones}')
print()
print(f'total de usuarios por {totales}')


total de conversiones por variante
control        779
tratamiento    820
Name: convirtio, dtype: int64

total de usuarios por variante
control        4965
tratamiento    5035
Name: convirtio, dtype: int64


In [165]:
conversion_resumen = experiment.groupby('variante')['convirtio'].mean()
print(f'Tasa de conversion por {conversion_resumen}')

Tasa de conversion por variante
control        0.156898
tratamiento    0.162860
Name: convirtio, dtype: float64


**Nota:** Se utiliza el metodo ***.sum()*** para sumar todos los valroes 1 de la columna ***convirtio***  para saber la cantidad numerica de conversión de los usuarios y ***.count()*** para ver el volumen completo de la muestra de usuarios. Con fines de evaluar la tasa de conversiones de ambas columnas se utiliza el metodo .***mean()***. 

In [166]:
from statsmodels.stats.proportion import proportions_ztest
z_stat, p_value = proportions_ztest(conversiones, totales)

print(f"Estadístico z: {z_stat}")
print(f"Valor p: {p_value}")

alpha = 0.05 
if p_value < alpha:
    print("Rechazamos la hipótesis nula: hay evidencia de una diferencia.")
else:
    print("No rechazamos la hipótesis nula: no hay evidencia suficiente de una diferencia.")

Estadístico z: -0.8132782986429474
Valor p: 0.41605851639119995
No rechazamos la hipótesis nula: no hay evidencia suficiente de una diferencia.


---

## 🔹 Paso 6: Comunicar los resultados (Dashboard en BI)

🎯 **Objetivo**:  
Crear un dashboard que muestre de manera clara y visual los resultados del análisis de ventas, costos, marketing y conversión. 

Se usarán los CSVs limpios del Paso 1:

- `orders_clean.csv`  
- `catalog_clean.csv`  
- `marketing_clean.csv`

---

1️⃣ Preparación de los datos
1. Cargar los CSVs en Power BI o Tableau.
2. Revisar relaciones:
   - `orders.nombre_producto` → `catalog.nombre_producto`
   - `orders.fecha_pedido` → tabla de fechas (crear calendario para análisis temporal)
   - `orders.fecha_pedido` → `dim_fecha.date`
3. Crear columnas calculadas necesarias
4. Crear tabla de fechas para poder calcular comparaciones YTD, YoY o períodos anteriores (`Previous Year`, `Previous Month`).

---

2️⃣ Dashboard 1: Overview Ejecutivo
**KPIs principales a mostrar:**
- Revenue total
- Profit total
- Gasto total en marketing
- Ticket promedio
- Cantidad promedio de productos por orden

**Visualizaciones sugeridas:**
- Tarjetas KPI para revenue, profit y gasto marketing
- Gráfico de líneas: evolución mensual de revenue o profit
- Gráfico de líneas YTD
- Gráfico de barras: revenue y profit por producto o categoría

---

 3️⃣ Dashboard 2: Detalle / Drill-through  
**Objetivo:** Permitir explorar los datos desde el KPI general hasta cada orden o producto.

**Visualizaciones sugeridas:**
- Tabla detallada de órdenes con:
  - producto, cantidad, revenue, cost, profit
  - color condicional (profit negativo en rojo, positivo en verde)
- Gráfico de barras por producto con medida `cantidad vendida`
- Drill-through: seleccionar un producto y ver todos los pedidos relacionados
- Filtros por fecha, categoría de producto, etc

---

## 🚀 Entrega Final

Comparte el acceso a tu Dashboard para revisión.   
Puedes entregar el Dashboard utilizando **Power BI o Tableau**.

Incluye **uno de los siguientes**:

- 🔗 Link público del dashboard publicado en **Power BI Service o Tableau Public / Tableau Cloud**
- 🔗 Link de **Google Drive o OneDrive** con el archivo del proyecto (`.pbix`) y los 3 csvs limpios.


### 📎 Enlace del Dashboard

In [167]:
# (Pega aquí tu link)
# link de power bi o tableau
# link de one drive / google drive

https://public.tableau.com/shared/4H2J6FSPM?:display_count=n&:origin=viz_share_link

https://drive.google.com/drive/folders/16BnpCladET8umQ3feZCKJW-vsJ5un409?usp=sharing